sale_exception_holidays_public/models/sale_order.py

```diff
@@ -0,0 +1,58 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import fields, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _is_commitment_date_a_public_holiday(self):
+        """
+        Returns True if commitment_date is a public holiday
+        :return: bool
+        """
+        self.ensure_one()
+        res = False
+        if not self.commitment_date:
+            return res
+        commitment_date = fields.Datetime.context_timestamp(
+            self, self.commitment_date
+        ).date()
+        holidays_filter = [("year", "=", commitment_date.year)]
+        partner = self.partner_shipping_id or self.partner_id
+        if partner:
+            if partner.country_id:
+                holidays_filter.append("|")
+                holidays_filter.append(("country_id", "=", False))
+                holidays_filter.append(("country_id", "=", partner.country_id.id))
+            else:
+                holidays_filter.append(("country_id", "=", False))
+        pholidays = self.env["hr.holidays.public"].search(holidays_filter)
+        if not pholidays:
+            return res
+
+        states_filter = [("year_id", "in", pholidays.ids)]
+        if partner.state_id:
+            states_filter += [
+                "|",
+                ("state_ids", "=", False),
+                ("state_ids", "=", partner.state_id.id),
+            ]
+        else:
+            states_filter.append(("state_ids", "=", False))
+        states_filter.append(("date", "=", commitment_date))
+        hhplo = self.env["hr.holidays.public.line"]
+        holidays_line = hhplo.search(states_filter, limit=1)
```

Comment:
domain = [
  ("year_id.country_id", "in", (False, partner.country_id.id)),
  ("state_ids", "in", (False, partner.state_id.id)),
  ("date", "=", commitment_date),
]
holidays_line = hhplo.search(domain, limit=1)

Reply:
@jbaudoux  Hi, i would like to keep it as it is. Because:
1. I copy the idea from [hr_holidays_public/models/hr_holidays_public.py#L77](
https://github.com/OCA/hr-holidays/blob/16.0/hr_holidays_public/models/hr_holidays_public.py#L77). i would like to respect the author's work.
2. "in" operator is against the indexes search, "=" operator is better.
3. relation fields: year_id.country_id,  state_ids in a single domain will lead to many sub queries which impacts the performance.

Reply:
I tried my suggested domain and odoo resolves that in 1 sql query. So your code is counter-optimized and over complexified for nothing. Odoo has improved a lot in query resolution and in this case, he resolves properly the domain.

Reply:
You're right, odoo handles your domain in a single query:
```
SELECT "hr_holidays_public_line".id FROM "hr_holidays_public_line" WHERE ((("hr_holidays_public_line"."year_id" in (SELECT "hr_holidays_public".id FROM "hr_holidays_public" WHERE (("hr_holidays_public"."country_id" in (233)) OR "hr_holidays_public"."country_id" IS NULL))) AND 
                        EXISTS (
                            SELECT 1 FROM "hr_holiday_public_state_rel" AS "hr_holidays_public_line__state_ids"
                            WHERE "hr_holidays_public_line__state_ids"."line_id" = "hr_holidays_public_line".id
                            AND "hr_holidays_public_line__state_ids"."state_id" IN (13)
                        )
                    ) AND ("hr_holidays_public_line"."date" = '2023-10-16')) ORDER BY  "hr_holidays_public_line"."date" ,"hr_holidays_public_line"."name" DESC  LIMIT 1 
```
But fyi, it makes the test failed, it does not handle the case `state_ids=False`, so i change your domain a bit:
```
        domain = [
            ("year_id.country_id", "in", (False, partner.country_id.id)),
            "|",
            ("state_ids", "=", False),
            ("state_ids", "=", partner.state_id.id),
            ("date", "=", commitment_date),
        ]
```
The query now looks like:
```
SELECT "hr_holidays_public_line".id FROM "hr_holidays_public_line" WHERE ((("hr_holidays_public_line"."year_id" in (SELECT "hr_holidays_public".id FROM "hr_holidays_public" WHERE (("hr_holidays_public"."country_id" in (233)) OR "hr_holidays_public"."country_id" IS NULL))) AND (
                        NOT EXISTS (
                            SELECT 1 FROM "hr_holiday_public_state_rel" AS "hr_holidays_public_line__state_ids"
                            WHERE "hr_holidays_public_line__state_ids"."line_id" = "hr_holidays_public_line".id
                        )
                     OR 
                        EXISTS (
                            SELECT 1 FROM "hr_holiday_public_state_rel" AS "hr_holidays_public_line__state_ids"
                            WHERE "hr_holidays_public_line__state_ids"."line_id" = "hr_holidays_public_line".id
                            AND "hr_holidays_public_line__state_ids"."state_id" IN (13)
                        )
                    )) AND ("hr_holidays_public_line"."date" = '2023-10-16')) ORDER BY  "hr_holidays_public_line"."date" ,"hr_holidays_public_line"."name" DESC  LIMIT 1 
```